### ISSUE TITLE
update _ids on has_many :through w/ conditions

### ISSUE THREAD

**Issue by codezomb** [rails/rails#18424]
Updating the association via the `_ids` method has some strange results.

When creating the association in this manner, the condition is correctly applied to set the `role` attribute.

``` sql
INSERT INTO "associations" ("associated_id", "associated_type", "created_at", "player_id", "role", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["associate d_id", 77], ["associated_type", "Message"], ["created_at", "2015-01-09 00:12:01.027972"], ["player_id", 66], ["role", "to"], ["updated_at", "2015-01-09 00:12:01.027972"]]
```

 However, when changing the attribute, `DELETE FROM` is called, but it's not scoped by the condition. This means that if you have two associations w/ different attributes (roles in this case), they both get deleted.

``` sql
DELETE FROM "associations" WHERE "associations"."associated_id" = $1 AND "associations"."associated_type" = $2 AND "associations"."player_id" = 66  [["associated_id", 77], ["associated_type", "Message"]]
```

Take the following code into account:

``` ruby
class Message < ActiveRecord::Base
  has_many :associations, as: :associated
  with_options through: :associations, source: :player do |assoc|
    assoc.has_many :from, -> { where(associations: { role: 'from' }) }
    assoc.has_many :to,   -> { where(associations: { role: 'to'   }) }
  end
end
```

``` ruby
class Association < ActiveRecord::Base
  belongs_to :associated, polymorphic: true
  belongs_to :player

  validates :role, presence: true
end
```

Now, the following example execution:

```
[145] pry(main)> Message.first.update_attributes(to_ids: [66], from_ids: [66])                                                                                                             │
  Message Load (0.5ms)  SELECT  "messages".* FROM "messages"   ORDER BY "messages"."id" ASC LIMIT 1
   (0.1ms)  BEGIN
  Player Load (0.3ms)  SELECT  "players".* FROM "players"  WHERE "players"."id" = $1  ORDER BY "players"."name" ASC LIMIT 1  [["id", 66]]
  Player Load (0.7ms)  SELECT "players".* FROM "players" INNER JOIN "associations" ON "players"."id" = "associations"."player_id" WHERE "associations"."role" = 'to' AND "associations"."associated_id" = $1 AND "associations"."associated_type" = $2  ORDER BY "players"."name" ASC  [["associated_id", 77], ["associated_type", "Message"]]
  SQL (0.2ms)  DELETE FROM "associations" WHERE "associations"."associated_id" = $1 AND "associations"."associated_type" = $2 AND "associations"."player_id" = 52  [["associated_id", 77], [ "associated_type", "Message"]]
  SQL (0.3ms)  INSERT INTO "associations" ("associated_id", "associated_type", "created_at", "player_id", "role", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["associate d_id", 77], ["associated_type", "Message"], ["created_at", "2015-01-09 00:12:01.027972"], ["player_id", 66], ["role", "to"], ["updated_at", "2015-01-09 00:12:01.027972"]]
  Player Load (0.1ms)  SELECT  "players".* FROM "players"  WHERE "players"."id" = $1  ORDER BY "players"."name" ASC LIMIT 1  [["id", 66]]
  Player Load (0.3ms)  SELECT "players".* FROM "players" INNER JOIN "associations" ON "players"."id" = "associations"."player_id" WHERE "associations"."role" = 'from' AND "associations"."a ssociated_id" = $1 AND "associations"."associated_type" = $2  ORDER BY "players"."name" ASC  [["associated_id", 77], ["associated_type", "Message"]]
  SQL (0.1ms)  INSERT INTO "associations" ("associated_id", "associated_type", "created_at", "player_id", "role", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["associated_id", 77], ["associated_type", "Message"], ["created_at", "2015-01-09 00:12:01.031852"], ["player_id", 66], ["role", "from"], ["updated_at", "2015-01-09 00:12:01.031852"]]
   (0.3ms)  COMMIT
=> true

[146] pry(main)> Message.first.update_attributes(to_ids: [66], from_ids: [])
  Message Load (0.4ms)  SELECT  "messages".* FROM "messages"   ORDER BY "messages"."id" ASC LIMIT 1
   (0.1ms)  BEGIN
  Player Load (0.1ms)  SELECT  "players".* FROM "players"  WHERE "players"."id" = $1  ORDER BY "players"."name" ASC LIMIT 1  [["id", 66]]
  Player Load (0.4ms)  SELECT "players".* FROM "players" INNER JOIN "associations" ON "players"."id" = "associations"."player_id" WHERE "associations"."role" = 'to' AND "associations"."associated_id" = $1 AND "associations"."associated_type" = $2  ORDER BY "players"."name" ASC  [["associated_id", 77], ["associated_type", "Message"]]
  Player Load (0.3ms)  SELECT "players".* FROM "players" INNER JOIN "associations" ON "players"."id" = "associations"."player_id" WHERE "associations"."role" = 'from' AND "associations"."associated_id" = $1 AND "associations"."associated_type" = $2  ORDER BY "players"."name" ASC  [["associated_id", 77], ["associated_type", "Message"]]
  SQL (0.2ms)  DELETE FROM "associations" WHERE "associations"."associated_id" = $1 AND "associations"."associated_type" = $2 AND "associations"."player_id" = 66  [["associated_id", 77], ["associated_type", "Message"]]
   (0.3ms)  COMMIT
=> true

[147] pry(main)> Message.first.to_ids
  Message Load (0.4ms)  SELECT  "messages".* FROM "messages"   ORDER BY "messages"."id" ASC LIMIT 1
   (0.3ms)  SELECT "players".id FROM "players" INNER JOIN "associations" ON "players"."id" = "associations"."player_id" WHERE "associations"."role" = 'to' AND "associations"."associated_id" = $1 AND "associations"."associated_type" = $2  ORDER BY "players"."name" ASC  [["associated_id", 77], ["associated_type", "Message"]]
=> []

[148] pry(main)> Message.first.from_ids
  Message Load (0.4ms)  SELECT  "messages".* FROM "messages"   ORDER BY "messages"."id" ASC LIMIT 1
   (0.3ms)  SELECT "players".id FROM "players" INNER JOIN "associations" ON "players"."id" = "associations"."player_id" WHERE "associations"."role" = 'from' AND "associations"."associated_id" = $1 AND "associations"."associated_type" = $2  ORDER BY "players"."name" ASC  [["associated_id", 77], ["associated_type", "Message"]]
=> []
```

**Comment by jamesmacwilliam** [comment]
:+1:

**Comment by bleiweiss** [comment]
:+1:

**Comment by rafaelfranca** [comment]
Could you create an reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

**Comment by Crunch09** [comment]
@rafaelfranca @codezomb  I could reproduce it with a failing test case https://gist.github.com/Crunch09/e08a8b3774f5bfd9b501 . If nobody has started yet, i'd be happy to try to write a fix for it.

**Comment by codezomb** [comment]
@Crunch09 thanks for providing the test case, I got tied up with deadlines.

**Comment by dmitry** [comment]
Still fails on edge:

```
BugTest#test_for_issue_18424 [a.rb:73]:
Expected: [1]
  Actual: []
```

**Comment by codezomb** [comment]
@kamipo This is still happening in 5.1.6. When I look for the changes in master it seems to have been reverted. Can we re-open?

**Comment by kamipo** [comment]
This should be fixed in 5.2.0.
